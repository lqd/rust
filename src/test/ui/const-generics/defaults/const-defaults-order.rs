// This test checks the diagnostics and ordering of various combinations of const generics defaults
// with other kinds of generic parameters used with and without bounds, with and without default
// values.

// revisions: min full
#![cfg_attr(full, feature(const_generics))]
#![cfg_attr(full, allow(incomplete_features))]
#![crate_type = "lib"]

// Under `min_const_generics`, the correct ordering is: lifetimes, then types, then consts, then
// defaulted parameters (both types and consts, with a default value). Additionally, no defaulted
// parameter can be used together with a const parameter.

// An example of good ordering, with incompatibilities: defaulted types used together with defaulted
// consts.
struct GoodOrdering<
    'a,
    'b: 'a,
    T,
    U: 'b,
    V = u8,
    W: 'b = i8,

    const O: usize = 0,
    //~^ ERROR default values for const generic parameters are experimental
    //[min]~^^ ERROR using type defaults and const parameters in the same parameter list

    const P: usize = 1
    //~^ ERROR default values for const generic parameters are experimental
    //[min]~^^ ERROR using type defaults and const parameters in the same parameter list
>(
    &'a (),
    &'b (),
    [T; O],
    [U; P],
    [V; O],
    [W; P],
);

// An example of bad ordering for every kind of parameter, with incompatibilities: defaulted types
// used together with defaulted consts.
struct BadOrdering<
    'a,
    T,
    const N: usize,

    V = u8,
    //[min]~^ ERROR using type defaults and const parameters in the same parameter list

    const O: usize = 0,
    //~^ ERROR default values for const generic parameters are experimental
    //[min]~^^ ERROR using type defaults and const parameters in the same parameter list

    'b: 'a,
    //~^ ERROR lifetime parameters must be declared prior to defaulted parameters

    W: 'b = i8,
    //[min]~^ ERROR using type defaults and const parameters in the same parameter list

    U: 'b,
    //~^ ERROR type parameters must be declared prior to defaulted parameters

    const P: usize = 1,
    //~^ ERROR default values for const generic parameters are experimental
    //[min]~^^ ERROR using type defaults and const parameters in the same parameter list

    const M: usize,
    //~^ ERROR const parameters must be declared prior to defaulted parameters
    //[min]~^^ ERROR using type defaults and const parameters in the same parameter list
>(
    &'a (),
    &'b (),
    [T; N],
    [U; M],
    [V; O],
    [W; P],
);
