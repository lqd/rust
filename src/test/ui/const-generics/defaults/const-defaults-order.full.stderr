error: lifetime parameters must be declared prior to defaulted parameters
  --> $DIR/const-defaults-order.rs:54:5
   |
LL |   struct BadOrdering<
   |  ___________________-
LL | |     'a,
LL | |     T,
LL | |     const N: usize,
...  |
LL | |     'b: 'a,
   | |     ^^
...  |
LL | |
LL | | >(
   | |_- help: reorder the parameters: `<'a, 'b: 'a, T, const N: usize, U: 'b, const M: usize, V = u8, const O: usize = 0, W: 'b = i8, const P: usize = 1>`
   |
   = note: the expected order for parameters is: lifetimes, then consts and types, then defaulted parameters

error: const parameters must be declared prior to defaulted parameters
  --> $DIR/const-defaults-order.rs:67:11
   |
LL |   struct BadOrdering<
   |  ___________________-
LL | |     'a,
LL | |     T,
LL | |     const N: usize,
...  |
LL | |     const M: usize,
   | |           ^
LL | |
LL | |
LL | | >(
   | |_- help: reorder the parameters: `<'a, 'b: 'a, T, const N: usize, U: 'b, const M: usize, V = u8, const O: usize = 0, W: 'b = i8, const P: usize = 1>`
   |
   = note: the expected order for parameters is: lifetimes, then consts and types, then defaulted parameters

error: type parameters must be declared prior to defaulted parameters
  --> $DIR/const-defaults-order.rs:60:5
   |
LL |   struct BadOrdering<
   |  ___________________-
LL | |     'a,
LL | |     T,
LL | |     const N: usize,
...  |
LL | |     U: 'b,
   | |     ^
...  |
LL | |
LL | | >(
   | |_- help: reorder the parameters: `<'a, 'b: 'a, T, const N: usize, U: 'b, const M: usize, V = u8, const O: usize = 0, W: 'b = i8, const P: usize = 1>`
   |
   = note: the expected order for parameters is: lifetimes, then consts and types, then defaulted parameters

error[E0658]: default values for const generic parameters are experimental
  --> $DIR/const-defaults-order.rs:24:20
   |
LL |     const O: usize = 0,
   |                    ^^^
   |
   = note: see issue #44580 <https://github.com/rust-lang/rust/issues/44580> for more information
   = help: add `#![feature(const_generics_defaults)]` to the crate attributes to enable

error[E0658]: default values for const generic parameters are experimental
  --> $DIR/const-defaults-order.rs:28:20
   |
LL |     const P: usize = 1
   |                    ^^^
   |
   = note: see issue #44580 <https://github.com/rust-lang/rust/issues/44580> for more information
   = help: add `#![feature(const_generics_defaults)]` to the crate attributes to enable

error[E0658]: default values for const generic parameters are experimental
  --> $DIR/const-defaults-order.rs:50:20
   |
LL |     const O: usize = 0,
   |                    ^^^
   |
   = note: see issue #44580 <https://github.com/rust-lang/rust/issues/44580> for more information
   = help: add `#![feature(const_generics_defaults)]` to the crate attributes to enable

error[E0658]: default values for const generic parameters are experimental
  --> $DIR/const-defaults-order.rs:63:20
   |
LL |     const P: usize = 1,
   |                    ^^^
   |
   = note: see issue #44580 <https://github.com/rust-lang/rust/issues/44580> for more information
   = help: add `#![feature(const_generics_defaults)]` to the crate attributes to enable

error: aborting due to 7 previous errors

For more information about this error, try `rustc --explain E0658`.
